common/devdoc/thandle_ptr_requirements.md

```diff
@@ -0,0 +1,91 @@
+# thandle_ptr_requirements
+
+## Overview
+
+`thandle_ptr` is a module that builds on `THANDLE` and provides a "move existing pointer under THANDLE" functionality.
+
+`THANDLE` has as requirement to "only call malloc once". `THANDLE_PTR` retains that requirement. Note: the existing pointer must have been allocated "somehow" before. Even pointers to stack variables can be moved under `THANDLE_PTR` with the regular gotchas that deallocation happens when leaving the scope of the variable.
+
+`THANDLE_PTR` should be used when the pointer already exists. In all the other situations, where the pointer hasn't been created yet, regular `THANDLE` should be used.
+
+`THANDLE_PTR` works by declaring under the hood a struct as below which captures the parameters passed to `THANDLE_PTR_CREATE_WITH_MOVE` and...
+
+```c
+typedef struct PTR_STRUCT_TAG_TYPE_NAME(T)      
+{                                               
+    T pointer;                                  
+    THANDLE_PTR_FREE_FUNC_TYPE_NAME(T) the_free;
+} PTR(T);
+```    
+
+... and then declaring a regular `THANDLE` over the above defined structure type.
+
+In this document and code implementation "T" is a previously defined type that has pointer semantics and is a C identifier. For example "PS" introduced as: `typedef struct S_TAG* PS;`
+
+## Usage
+
+Usage of `THANDLE_PTR` is very similar to `THANDLE`. 
+
+In a C header, introduce the following declaration:
+```c
+THANDLE_PTR_DECLARE(A_S_PTR);
+```
+
+In a C source file, introduce the following declaration:
+```c
+THANDLE_PTR_DEFINE(A_S_PTR);
+```
+
+Then use it like:
+```c
+THANDLE(PTR(A_S_PTR)) one = THANDLE_PTR_CREATE_WITH_MOVE(A_S_PTR)(a_s, dispose);
+```
+
+`a_s` is a pointer of type `A_S_PTR`.
+
+The above code "moves" ("captures") the pointer `a_s` and produces the `THANDLE` `one`. `dispose` is a user function that takes the original `a_s` pointer and frees all resource used by `a_s`.
+
+`one->pointer` produces the original `a_s` pointer. Other fields of the `one` pointer should not be used (it also captures the `dispose` function for example).
```

Comment:
>one->pointer [](http://example.com/codeflow?start=1&length=12)

Should the access to this pointer be made a macro as well?

Reply:
The name of the parameter passed to CREATE is `pointer` so maaaaybe this is good enough as it is. I think usage will tell if enough or not.

Reply:
However, this "because it is a THANDLE", the GET_T of the THANDLE should get the pointer out.

... with->pointer I am afraid the *pointer is read only. Give me 1 second to write another test :)

Reply:
I wrote some more tests where I prove that the ->pointer's fields have the same qualifiers as the original one.